File: 0680_valid_palindrome_ii/solution.go
// LeetCode 0680 - Valid Palindrome II
// https://leetcode.com/problems/valid-palindrome-ii/

func validPalindrome(s string) bool {
	isPalindrome := func(left, right int) bool {
		for left < right {
			if s[left] != s[right] {
				return false
			}
			left++
			right--
		}
		return true
	}
	left, right := 0, len(s)-1
	for left < right {
		if s[left] != s[right] {
			return isPalindrome(left+1, right) || isPalindrome(left, right-1)
		}
		left++
		right--
	}
	return true
}

File: 0680_valid_palindrome_ii/solution.rb
# LeetCode 0680 - Valid Palindrome II
# https://leetcode.com/problems/valid-palindrome-ii/

# @param {String} s
# @return {Boolean}
def valid_palindrome(s)
  is_palindrome = lambda do |left, right|
    while left < right
      return false if s[left] != s[right]

      left += 1
      right -= 1
    end
    true
  end

  left = 0
  right = s.length - 1
  while left < right
    if s[left] != s[right]
      return is_palindrome.call(left + 1, right) || is_palindrome.call(left, right - 1)
    end

    left += 1
    right -= 1
  end
  true
end
